### PR TITLE
plugin/kubernetes: Expose Load Balancer/External IPs as records

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

Adds option to expose Services with external Load Balancer IPs (and External IPs) in the form:

`service-name.namespace.zone`

where zone is a zone defined as an external zone in the options.  e.g.

```
kubernetes cluster.local myzone.com in-addr.arpa ip6.arpa {
    external myzone.com
}
```

### 2. Which issues (if any) are related?
#1851
kubernetes/dns#242
#2349

### 3. Which documentation changes (if any) need to be made?

### CODE REVIEW

**Comment by chrisohaver** [comment]
Bulk of functionality is working...

Still to do:
- ~~zone transfers~~
- ~~unit tests~~
- ~~ci tests~~

File: plugin/kubernetes/controller.go [hunk 1]
```diff
@@ -167,14 +181,30 @@ func podIPIndexFunc(obj interface{}) ([]string, error) {
 	return []string{p.PodIP}, nil
 }
 
-func svcIPIndexFunc(obj interface{}) ([]string, error) {
+func svcAllIPsIndexFunc(obj interface{}) ([]string, error) {
+	svc, ok := obj.(*object.Service)
+	if !ok {
+		return nil, errObj
+	}
+	return append(svc.ExternalIPs, svc.ClusterIP), nil
```
Comment (johnbelamaric) [inline plugin/kubernetes/controller.go:189]: How does the indexing work? Does this create two index entries, one for each IP?

File: plugin/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -110,12 +118,16 @@ func (k *Kubernetes) Services(state request.Request, exact bool, opt plugin.Opti
 
 	case dns.TypeNS:
 		// We can only get here if the qname equals the zone, see ServeDNS in handler.go.
+		if external {
+			svc := msg.Service{Host: state.LocalIP(), Key: msg.Path(state.QName(), "coredns")}
```
Comment (johnbelamaric) [inline plugin/kubernetes/kubernetes.go]: so, this is interesting. what does the NS record look like for the external zone?

File: plugin/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -110,12 +118,16 @@ func (k *Kubernetes) Services(state request.Request, exact bool, opt plugin.Opti
...
+			return []msg.Service{svc}, nil
+		}
 		ns := k.nsAddr()
 		svc := msg.Service{Host: ns.A.String(), Key: msg.Path(state.QName(), "coredns")}
 		return []msg.Service{svc}, nil
 	}
 
-	if state.QType() == dns.TypeA && isDefaultNS(state.Name(), state.Zone) {
+	if state.QType() == dns.TypeA && isDefaultNS(state.Name(), state.Zone) && !external {
```
Comment (johnbelamaric) [inline plugin/kubernetes/kubernetes.go]: do we need some way to get the external IP for CoreDNS? I think we should set this up to support delegation. I think that means that we don't necessarily need that, because the parent zone file should contain the A record for the NS in the delegation. @miekg or @cricketliu can you clarify how that would work? For example, if the external zone is "cluster.example.com":

CoreDNS returns SOA for external domain:

```
cluster.example.com. IN SOA ns.dns.cluster.example.com hostmaster.cluster.example.com (...)
```

example.com contains

```
cluster.example.com. IN NS ns.dns.cluster.example.com
ns.dns.cluster.example.com IN A <coredns-service-external-ip>
```

Does CoreDNS need to have an A record for ns.dns.cluster.example.com, or is the glue record in example.com sufficient?

**Review by johnbelamaric — COMMENTED** [review]
This will be very cool. A few comments. One question: here, we do this for load balancer services. But what about other ways to expose services? See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#servicespec-v1-core

In particular, I am thinking about:

- externalIPs
- SRV records for NodePort services

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:189]: Yes

File: plugin/kubernetes/kubernetes.go (hunk 2 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: `ns.dns.zone`

Comment (johnbelamaric) [reply plugin/kubernetes/kubernetes.go]: Ok. but does it also create an A record? I don't get what LocalIP() is doing here?

**Comment by chrisohaver** [comment]
> In particular, I am thinking about:
> * externalIPs

Regarding "External IPs",  I'm confused about the the exact relationship with load balancers ips... the docs say that load balancers are an example of external ips... are they the only example?  When I create load balancers in k8s, external IPs field remains blank on the service.  So I was thinking External IPs field may be used by some cloud providers to allow the user to declare the LB IP instead of it being dynamically provisioned.  The only examples of "external ips" I could find involve load balancers.  if you could point me to a non loadbalancer external IP example that would help tremendously.

 > * SRV records for NodePort services

That would be cool. I think it should be delivered in a separate PR.

**Comment by johnbelamaric** [comment]
Agree on the SRVs for NodePort - separate PR is best.

For externalIPs, they are just managed outside of Kubernetes. So, the load
balancer ones are managed by the cloud controller (or plugin), but these
are just specified by the user. The user has to manually configure any
relevant nodes to accept traffic on those IPs. They also need to control
the pod placement to land on one of those nodes. SRV probably makes sense
here as well, because you may have several. Though, depending on your
routing (Anycast, for example), you could use the same IP on multiple nodes.

If we choose to do this one it should be a separate PR as well. Probably
need more real-world examples.

File: plugin/kubernetes/kubernetes.go (hunk 3 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: We have code already to get the cluster IP of coredns in cluster.
If we need pull the external ip, I can adapt that code.
Of course, that assuming the in-cluster coreDNS _has_ an external ip... which is does not by default.

I was thinking that for most cases, if someone wants to expose external records, they'd do so in an external instance of coredns, not serve them from the in-cluster replicas... but i could be wrong.

**Comment by chrisohaver** [comment]
> For externalIPs, they are just managed outside of Kubernetes. ...The user has to manually configure any relevant nodes to accept traffic on those IPs.

OK, makes sense.  In that case it's very easy to add.  I will do that in this PR, and update docs.

File: plugin/kubernetes/kubernetes.go (hunk 2 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: yes, it creates an A record too...  I thought it was supposed to be the ip address of coredns itself.

for the in-cluster zone, this is the ip address of the coredns service (e.g. `10.96.0.10`).  so you have (i think)...
```
ns.dns.cluster.local IN NS kube-dns.kube-system.svc.cluster.local
kube-dns.kube-system.svc.cluster.local IN A 10.96.0.10
```

For the external zone, i think it should be something like ...
```
ns.dns.cluster.myzone.com IN NS kube-dns.kube-system.myzone.com
kube-dns.kube-system.myzone.com IN A 1.2.3.4
```

Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: isn't this the same question as below? "do we need some way to get the external IP for CoreDNS?"

File: plugin/kubernetes/kubernetes.go (hunk 3 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: Looking a second time, I see some problems with the in-cluster kubernetes NS record code that need fixing ...

Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: Also, for external zones, `ns.dns.zone.` collides with a service named `ns` in namespace `dns`.

This whole mess is related to the backend hard coding `ns.dns.zone.` as the nameserver portion of an `SOA`  record.

**Comment by miekg** [comment]
It needs to be somewhere..any suggestions?

**Comment by chrisohaver** [comment]
> It needs to be somewhere..any suggestions?

I think in the case of kubernetes, it could point directly to the dns service... `kube-dns.kube-system.zone.`

```
zone. IN NS kube-dns.kube-system.zone.
kube-dns.kube-system.zone. IN A 1.2.3.4
```
This means the backend would need to get the NS name is from the plugin when generating an `SOA` ...

**Comment by miekg** [comment]
We could easily change that and use the kube-system name space

**Comment by cricketliu** [comment]
On Nov 30, 2018, at 12:01 PM, John Belamaric <notifications@github.com<mailto:notifications@github.com>> wrote:


do we need some way to get the external IP for CoreDNS? I think we should set this up to support delegation. I think that means that we don't necessarily need that, because the parent zone file should contain the A record for the NS in the delegation. @miekg<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_miekg&d=DwMCaQ&c=CWsWqoUynJrLESJQduOsxQ&r=uA3zJBZARmPCCWmTYBLk07UikWEyE9AKvAmI_Idoujg&m=WKiJZ8IOBtMAOh043-2de8vJoRzylFalwrXka2IK18s&s=LwRlYFcsgbHhWg8s5KdASaVmRmzaI_BjGsUKVm9ptTs&e=> or @cricketliu<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_cricketliu&d=DwMCaQ&c=CWsWqoUynJrLESJQduOsxQ&r=uA3zJBZARmPCCWmTYBLk07UikWEyE9AKvAmI_Idoujg&m=WKiJZ8IOBtMAOh043-2de8vJoRzylFalwrXka2IK18s&s=mxrOv80aG2GU_J9k-pW8yEVfstE76AwJWcM_UJiZicc&e=> can you clarify how that would work? For example, if the external zone is "cluster.example.com<http://cluster.example.com/>":

CoreDNS returns SOA for external domain:

cluster.example.com<http://cluster.example.com/>. IN SOA ns.dns.cluster.example.com<http://ns.dns.cluster.example.com/> hostmaster.cluster.example.com<http://hostmaster.cluster.example.com/> (...)


example.com<http://example.com/> contains

cluster.example.com<http://cluster.example.com/>. IN NS ns.dns.cluster.example.com<http://ns.dns.cluster.example.com/>
ns.dns.cluster.example.com<http://ns.dns.cluster.example.com/> IN A <coredns-service-external-ip>


Does CoreDNS need to have an A record for ns.dns.cluster.example.com<http://ns.dns.cluster.example.com/>, or is the glue record in example.com<http://example.com/> sufficient?

If CoreDNS is authoritative for cluster.example.com<http://cluster.example.com>, and ns.dns.cluster.example.com<http://ns.dns.cluster.example.com> is in that zone, then yes, CoreDNS needs an A RR for ns.dns.cluster.example.com<http://ns.dns.cluster.example.com>, too.  In other words, the A RR needs to appear twice:  Once in example.com<http://example.com> as glue, and once in the authoritative zone it's a part of.

cricket

**Comment by cricketliu** [comment]
On Nov 30, 2018, at 12:36 PM, chrisohaver <notifications@github.com<mailto:notifications@github.com>> wrote:


yes, it creates an A record too... I thought it was supposed to be the ip address of coredns itself.

for the in-cluster zone, this is the ip address of the coredns service (e.g. 10.96.0.10). so you have (i think)...

ns.dns.cluster.local IN NS kube-dns.kube-system.svc.cluster.local
kube-dns.kube-system.svc.cluster.local IN A 10.96.0.10


For the external zone, i think it should be something like ...

ns.dns.cluster.myzone.com<http://ns.dns.cluster.myzone.com/> IN NS kube-dns.kube-system.myzone.com<http://kube-dns.kube-system.myzone.com/>
kube-dns.kube-system.myzone.com<http://kube-dns.kube-system.myzone.com/> IN A 1.2.3.4


We should probably test whether or not the A RR is necessary to complete the delegation.  The test is straightforward:  An A RR is needed as "glue" iff the name of the name server ends in the name of the subdomain.   So in this case, kube-dns.kube-system.myzone.com<http://kube-dns.kube-system.myzone.com> doesn't end in ns.dns.cluster.myzone.com<http://ns.dns.cluster.myzone.com>, so no A RR is needed.

cricket

File: plugin/kubernetes/README.md [hunk 4]
```diff
@@ -45,6 +45,7 @@ kubernetes [ZONES...] {
     transfer to ADDRESS...
     fallthrough [ZONES...]
     ignore empty_service
+    external ZONES...
```
Comment (miekg) [inline plugin/kubernetes/README.md:48]: is ZONES optional?

File: plugin/kubernetes/README.md [hunk 5]
```diff
@@ -152,6 +157,27 @@ kubernetes cluster.local {
 }
 ~~~
 
+## Exposing Load Balancer IPs
+
+Expose internal service cluster ips in the `cluster.local` zone, and external load
```
Comment (miekg) [inline plugin/kubernetes/README.md]: s/ips/IPs/g

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (miekg) [reply plugin/kubernetes/controller.go:189]: why is this an append?

File: plugin/kubernetes/kubernetes.go (hunk 2 above)
Comment (miekg) [reply plugin/kubernetes/kubernetes.go]: yes, we need to think about it (probably in a new issue to not muddle this).

File: plugin/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -274,11 +289,24 @@ func (k *Kubernetes) InitKubeCache() (err error) {
 
 // Records looks up services in kubernetes.
 func (k *Kubernetes) Records(state request.Request, exact bool) ([]msg.Service, error) {
-	r, e := parseRequest(state)
+
+	external := k.externalZone(state.Name())
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: externalZones is a 1 -line function just inline it here (and the other place you call it)

File: plugin/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -274,11 +289,24 @@ func (k *Kubernetes) InitKubeCache() (err error) {
...
+
+	var (
+		r recordRequest
+		e error
+	)
+
+	if external {
+		r, e = parseRequestExternal(state)
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: Why does this need an almost identical parse function?

File: plugin/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -290,12 +318,12 @@ func (k *Kubernetes) Records(state request.Request, exact bool) ([]msg.Service,
...
 	}
 
-	if r.podOrSvc == Pod {
+	if r.podOrSvc == Pod && !external {
 		pods, err := k.findPods(r, state.Zone)
 		return pods, err
 	}
 
-	services, err := k.findServices(r, state.Zone)
+	services, err := k.findServices(r, state.Zone, external)
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: The external service search might actually be cleaner in a separate function.

File: plugin/kubernetes/object/service.go [hunk 9]
```diff
@@ -12,6 +12,7 @@ type Service struct {
 	Namespace    string
 	Index        string
 	ClusterIP    string
+	ExternalIPs  []string
```
Comment (miekg) [inline plugin/kubernetes/object/service.go:15]: this adds extra memory, have we tested it?

File: plugin/kubernetes/setup.go [hunk 10]
```diff
@@ -287,11 +275,68 @@ func ParseStanza(c *caddy.Controller) (*Kubernetes, error) {
 				continue
 			}
 			return nil, c.ArgErr()
+		case "external":
+			args := c.RemainingArgs()
+			if len(args) > 0 {
```
Comment (miekg) [inline plugin/kubernetes/setup.go]: please return on len(args) == 0 and dedent the rest of this block

File: plugin/pkg/dnsutil/reverse.go [hunk 11]
```diff
@@ -39,6 +39,12 @@ func IsReverse(name string) int {
 	if strings.HasSuffix(name, IP6arpa) {
 		return 2
 	}
+	if "."+name == IP4arpa {
```
Comment (miekg) [inline plugin/pkg/dnsutil/reverse.go]: this creates garbage.

also why?

File: plugin/kubernetes/setup.go [hunk 12]
```diff
@@ -287,11 +275,68 @@ func ParseStanza(c *caddy.Controller) (*Kubernetes, error) {
...
+						if dnsutil.IsReverse(z) > 0 {
+							continue
+						}
+						if z == extZone {
+							found = true
+							break
+						}
+					}
+					if !found {
+						return nil, c.Errf("external zone '%s' is not in plugin/block zone list", c.Val())
```
Comment (miekg) [inline plugin/kubernetes/setup.go]: none of the plugins have these checks, I'm also fan of garbage in, garbage out, so we should add it here.
As the corefile has all context we can easily write a tools that checks the config, ala named-checkconfig instead of pulling out that code directly into coredns

**Review by miekg — CHANGES_REQUESTED** [review]

File: plugin/kubernetes/README.md (hunk 4 above)
Comment (chrisohaver) [reply plugin/kubernetes/README.md:48]: No, Zones is not optional.  At least one is required.
You're basically defining which zones will be treated as external zones.

**Comment by chrisohaver** [comment]
Cant reply to original code comment for some buggy reason.

> why is this an append?

Because i want to append a record to a slice.  Is there a better way?

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:189]: Because i want to append a record to a slice.  Is there a better way?

File: plugin/kubernetes/kubernetes.go (hunk 7 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: I felt it was better than cluttering the code with IFs...

File: plugin/kubernetes/kubernetes.go (hunk 8 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: That would result in some code duplication.  You OK with that?

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (chrisohaver) [reply plugin/kubernetes/object/service.go:15]: No, I have not tested it yet. I don't expect it will add much.

File: plugin/pkg/dnsutil/reverse.go (hunk 11 above)
Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: As it was, `IsReverse("in-addr.arpa.")` would return `0`, because it does not have the suffix ".in-addr.arpa.".

I'll try to fix this differently.

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (johnbelamaric) [reply plugin/kubernetes/object/service.go:15]: If it's not too hard we can not keep this if no external zone is set. But maybe we don't know that this deep in.

Comment (chrisohaver) [reply plugin/kubernetes/object/service.go:15]: What is the memory footprint for an empty slice?

**Comment by johnbelamaric** [comment]
It's a pointer. So I would think 8 bytes typically.

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (chrisohaver) [reply plugin/kubernetes/object/service.go:15]: For most services, this will be empty.  I think the impact will be very minimal.  IMO, we can optimize later if we need to.

File: plugin/kubernetes/README.md [hunk 13]
```diff
@@ -109,6 +110,10 @@ kubernetes [ZONES...] {
 * `ignore empty_service` return NXDOMAIN for services without any ready endpoint addresses (e.g., ready pods).
   This allows the querying pod to continue searching for the service in the search path.
   The search path could, for example, include another Kubernetes cluster.
+* `external` will synthesize records in the listed **[ZONES...]** for all Kubernetes services with Load Balancer IPs.
```
Comment (miekg) [inline plugin/kubernetes/README.md]: which ZONES is this referencing?

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (miekg) [reply plugin/kubernetes/controller.go:189]: oh, non obvious, can this be done in the caller or don't we control that?

File: plugin/kubernetes/kubernetes.go [hunk 14]
```diff
@@ -110,21 +117,30 @@ func (k *Kubernetes) Services(state request.Request, exact bool, opt plugin.Opti
...
-		// If this is an A request for "ns.dns", respond with a "fake" record for coredns.
-		// SOA records always use this hardcoded name
-		ns := k.nsAddr()
-		svc := msg.Service{Host: ns.A.String(), Key: msg.Path(state.QName(), "coredns")}
-		return []msg.Service{svc}, nil
+		// create an A record for the NS target
+		nss := k.nsAddr(external)
+		var svcs []msg.Service
+		for _, ns := range nss {
+			if ns.A.String() == "0.0.0.0" {
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: what?

File: plugin/kubernetes/parse.go [hunk 15]
```diff
@@ -57,13 +58,16 @@ func parseRequest(state request.Request) (r recordRequest, err error) {
...
 	}
-	r.podOrSvc = segs[last]
-	if r.podOrSvc != Pod && r.podOrSvc != Svc {
-		return r, errInvalidRequest
-	}
-	last--
-	if last < 0 {
-		return r, nil
+
+	if !external {
```
Comment (miekg) [inline plugin/kubernetes/parse.go:64]: why can't the caller check the return value of this function as-is and then decide if the request is usable. Why does this function need changing?

File: plugin/pkg/dnsutil/reverse.go [hunk 16]
```diff
@@ -33,12 +33,13 @@ func ExtractAddressFromReverse(reverseName string) string {
 // name is in a reverse zone. The returned integer will be 1 for in-addr.arpa. (IPv4)
 // and 2 for ip6.arpa. (IPv6).
 func IsReverse(name string) int {
-	if strings.HasSuffix(name, IP4arpa) {
+	if strings.HasSuffix(name, IP4arpa) || name == IP4arpaDom {
 		return 1
 	}
-	if strings.HasSuffix(name, IP6arpa) {
+	if strings.HasSuffix(name, IP6arpa) || name == IP6arpaDom {
```
Comment (miekg) [inline plugin/pkg/dnsutil/reverse.go]: this should be needed.

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:189]: I don't think so, this is an indexer func thats called by the k8s client when k8s Services are changed/added/deleted, it expects a single slice of strings. 

An alternative would be to maintain a separate index.  But there's more overhead with a maintaining an additional index.

File: plugin/pkg/dnsutil/reverse.go (hunk 16 above)
Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: ?

File: plugin/kubernetes/parse.go (hunk 15 above)
Comment (chrisohaver) [reply plugin/kubernetes/parse.go:64]: Because the request format is different for external vs cluster names.

external names: service-name.namespace.zone.
cluster names: service-name.namespace.svc.zone.

Also, cluster names accept an endpoint segment for A records of service endpoints (pod ips). (endpoint.service-name.namespace.svc.zone).  External names do not (they only expose external IPs).

File: plugin/kubernetes/README.md (hunk 13 above)
Comment (chrisohaver) [reply plugin/kubernetes/README.md]: The local one in the option.  I'll clarify...

File: plugin/pkg/dnsutil/reverse.go (hunk 16 above)
Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: Do you mean, "This should _not_ bet needed." ?

This does fix a bug in the existing code, where `in-addr.arpa.` is not identified as a reverse domain.
This can cause the _kubernetes_ plugin to select a reverse domain as its primary domain if `in-addr.arpa` is listed before `cluster.local`. e.g. `kubernetes in-addr.arpa cluster.local`

File: plugin/kubernetes/kubernetes.go (hunk 14 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: Yeah - I should have commented the code here. I am using `0.0.0.0` to mean a null IP (i.e. we don't know the IP). I will refactor the nsAddr() function to make this less ugly.

File: plugin/kubernetes/parse.go (hunk 15 above)
Comment (miekg) [reply plugin/kubernetes/parse.go:64]: ah, that would be a good comment to put here.

File: plugin/pkg/dnsutil/reverse.go (hunk 16 above)
Comment (miekg) [reply plugin/pkg/dnsutil/reverse.go]: yeah, missed *not*, just spotted that.

in-addr.arpa can never be a correct reverse domain in our context.
It most def. should *not* qualify as "true" in `IsReverse`

Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: Hmm. OK.  I'll create a different function that does what I need, and remove IsReverse(), since it will no longer be used by anything.

Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: ... actually - i'll create a new function that only checks the domain, and call both.

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:189]:
```
return []string{svc.ExternalIPs, svc.ClusterIP}, nil
```
Is that any better?

Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:189]: oh, nevermind. I see svc.ExternalIPs is already a slice

File: plugin/pkg/dnsutil/reverse.go (hunk 16 above)
Comment (chrisohaver) [reply plugin/pkg/dnsutil/reverse.go]: Ive added a new function, and also call that when i'm comparing zones that may be a base reverse domain.

File: plugin/kubernetes/kubernetes.go (hunk 14 above)
Comment (chrisohaver) [reply plugin/kubernetes/kubernetes.go]: I added comments, and use nil instead, which is a lot clearer.

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (miekg) [reply plugin/kubernetes/object/service.go:15]: https://play.golang.org/p/TB_uzDhcwXc
12 bytes, this has to be moved somewhere else

Comment (chrisohaver) [reply plugin/kubernetes/object/service.go:15]: Well, with 10,000 services max, 12 bytes per service comes to 0.12 MB
Which comes to about a 0.06% increase when comparing to scale results for clusters with ~8600 services (approx 189 MB total mb footprint + 0.12 MB).

Given that, I don't think this is a significant enough of an increase to warrant moving it.

Comment (chrisohaver) [reply plugin/kubernetes/object/service.go:15]: > with 10,000 services max ...

To clarify, 10,000 services is the maximum number of services allowed in a single Kubernetes cluster (the entire cluster. not per node).

If we consider the worst possible case (not realistic, but technically possible): a single node cluster with the maximum 10,000 services, the proportional increase due to empty 12 bytes slices would be about 0.21% (approx 58 MB total mb footprint + 0.12 MB).

**Comment by chrisohaver** [comment]
Regarding the `NS` records - I have added ci integration tests for them for local and external zones.

The current behavior is to respond with an `NS` record for `ns.dns.<zone>` and also include `A` records in the additional section.  

- For the cluster's local zone, the `A` record is for the Cluster IP of the CoreDNS Service. 
- For the external zone, the additional section includes an `A` record for each External IP of the CoreDNS Service.

The CoreDNS Service is determined by looking the local IP in the Endpoint object store.

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (miekg) [reply plugin/kubernetes/object/service.go:15]: your almost doubling the default size  of this struct. And given that we had a lengthy  process of getting the memory down we should not add stuff here.

**Comment by johnbelamaric** [comment]
Can we run a test to see how much of an effect this has? It's an empirical
question...

**Comment by chrisohaver** [comment]
> Can we run a test to see how much of an effect this has?

Yes, I'll get one started today.

**Comment by chrisohaver** [comment]
>>Can we run a test to see how much of an effect this has?

>Yes, I'll get one started today.

The 500 node test result shows a 1% increase in memory usage.
(compared to CoreDNS 1.2.4)

**Comment by miekg** [comment]
regarding parseQuery: for a name like `service-name.namespace.zone` you want to know 2 things after you've established it's an external zone
1. the namespace (i.e. first label before the zone)
2. service-name, (i.e. the label before that)

So parse query should not have to be changed it all to make this work

**Comment by miekg** [comment]
> The 500 node test result shows a 1% increase in memory usage.
> (compared to CoreDNS 1.2.4)

Good to know!

How hard is it to *not* change the Service struct?

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (miekg) [reply plugin/kubernetes/controller.go:189]: the append's return value is not assigned to anything

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (miekg) [reply plugin/kubernetes/object/service.go:15]: `ExternalIPs:  make([]string, len(svc.Status.LoadBalancer.Ingress)+len(svc.Spec.ExternalIPs))`
will probably somewhat more than 12 bytes

File: plugin/kubernetes/object/service.go [hunk 17]
```diff
@@ -47,6 +49,16 @@ func ToService(obj interface{}) interface{} {
 		copy(s.Ports, svc.Spec.Ports)
 	}
 
+	for i, lb := range svc.Status.LoadBalancer.Ingress {
+		s.ExternalIPs[i] = lb.IP
+	}
+
+	ipCount := len(svc.Status.LoadBalancer.Ingress)
```
Comment (miekg) [inline plugin/kubernetes/object/service.go:56]: is `i` of the same value as you just iterate over the slice?

**Comment by chrisohaver** [comment]
> regarding parseQuery: for a name like service-name.namespace.zone you want to know 2 things after you've established it's an external zone
>the namespace (i.e. first label before the zone)
>service-name, (i.e. the label before that)
>So parse query should not have to be changed it all to make this work

the unmodified parse function would puke on "namespace" because it is not "svc" or "pod"...

**Comment by chrisohaver** [comment]
> How hard is it to not change the Service struct?

We could refactor to make all object.Service access through an interface, and make two structs that implement the interface (one with externalIPs and one without).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/kubern..." ]
>> How hard is it to not change the Service struct?
>
>We could refactor to make all object.Service access through an interface, and make two structs that implement the interface (one with externalIPs and one without).

hmmm, that would make things slower and can also increase memory.

I still would like to avoid changing this struct, unless it's a feature that 
benefits everyone. Right now every pays the price for a minority feature.

/Miek

--
Miek Gieben

**Comment by chrisohaver** [comment]
FYI, @johnbelamaric, this wont be done by kube-con. Sorry!

File: plugin/kubernetes/controller.go (hunk 1 above)
Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:189]: ? it's returned to the caller which is inside the client-go code.

File: plugin/kubernetes/object/service.go (hunk 9 above)
Comment (johnbelamaric) [reply plugin/kubernetes/object/service.go:15]: Well, for most services the becomes make([]string, 0)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/kubern..." ]
>johnbelamaric commented on this pull request.
>
>
>
>> @@ -167,14 +181,30 @@ func podIPIndexFunc(obj interface{}) ([]string, error) {
> 	return []string{p.PodIP}, nil
> }
>
>-func svcIPIndexFunc(obj interface{}) ([]string, error) {
>+func svcAllIPsIndexFunc(obj interface{}) ([]string, error) {
>+	svc, ok := obj.(*object.Service)
>+	if !ok {
>+		return nil, errObj
>+	}
>+	return append(svc.ExternalIPs, svc.ClusterIP), nil
>
>? it's returned to the caller which is inside the client-go code.
>

no other function does this in this file, why does this one need to do it?

**Comment by johnbelamaric** [comment]
The other index funcs return a slice with a single element, or in the case
of the external IP one, the existing slice. In this case we want to index
by all the IPs, and we don't know the exact number (and it will be
different for each service).

@chrisohaver is this redundant with the separate cluster IP and external ip
indexes?

**Comment by johnbelamaric** [comment]
I think this will be a pretty popular feature. We could make a separate
ToService function that doesn’t store the external ips, and choose which
one to use based on the setting. This means if you are not using the
feature, ExternalIPs field can always be nil.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/kubern..." ]
>I think this will be a pretty popular feature. We could make a separate
>ToService function that doesn’t store the external ips, and choose which
>one to use based on the setting. This means if you are not using the
>feature, ExternalIPs field can always be nil.

OTOH the 1% might not be much, but it's more than nothing and where to we draw 
the line? We were really counting the percentages back when.

**Comment by chrisohaver** [comment]
Yes, we can easily do that. Though I don’t believe any services have external ips in the k8s cluster scale tests, so that test result would not be affected.

We have about 25% headroom under the memory limit at highest scale test.

On Dec 6, 2018, at 5:13 PM, John Belamaric <notifications@github.com<mailto:notifications@github.com>> wrote:

I think this will be a pretty popular feature. We could make a separate
ToService function that doesn’t store the external ips, and choose which
one to use based on the setting. This means if you are not using the
feature, ExternalIPs field can always be nil.

**Comment by miekg** [comment]
...compared to kubedns. That 25% is not headroom, it's us being more efficient. In fact that scale test should slash 25% of that target and give those bytes to other processes.

Another thing this PR does is allowing a plugin to add zones to the initial list, no other plugin does this and changes the semantics of the Corefile considerably

**Comment by chrisohaver** [comment]
Let me rephrase. We could grow 25% before hitting the coredns deployment’s memory limit.

This option does not add zones. The external zone must already be in the initial list.

On Dec 7, 2018, at 1:18 AM, Miek Gieben <notifications@github.com<mailto:notifications@github.com>> wrote:


...compared to kubedns. That 25% is not headroom, it's us being more efficient. In fact that scale test should slash 25% of that target and give those bytes to other processes.

Another thing this PR does is allowing a plugin to add zones to the initial list, no other plugin does this and changes the semantics of the Corefile considerably

File: plugin/backend_lookup.go [hunk 18]
```diff
@@ -374,6 +374,7 @@ func NS(b ServiceBackend, zone string, state request.Request, opt Options) (reco
 	// ... and reset
 	state.Req.Question[0].Name = old
 
+	hostSeen := make(map[string]bool)
```
Comment (miekg) [inline plugin/backend_lookup.go:377]: sure, another expensive map in the serving path, why not?

**Comment by miekg** [comment]
Let me give a list on why this PR will not be merged. TL;DR: it increasing the maintenance burden in the future.

* it's a tiny feature
* functions get magical new boolean arguments that subtlety changes how they work.
* the hot serving path needs to be changed
* good documention in the sources is lacking

Good things:

* tests are good

Some other things that didn't help:

* putting an artificial timeframe on it (this has to be merged before xxxx)
* scaring us that its either this or a 3000 line copy/pasta PR (current alternative PR is ~500 lines minus tests)
